archive/100-theorems-list/70_perfect_numbers.lean

```diff
@@ -0,0 +1,671 @@
+    { intro z, simp only [finset.mem_insert, finset.mem_singleton], intro h,
+      cases h; rw h, apply hdvd } },
+  have card2 := card_pair_eq_2 hneq, rw hdivs at card2, rw ← prime_iff_two_divisors at card2,
+  split, apply card2,
+  rw divisors_prime card2 at hdivs,
+  have memy : y ∈ ({x, y} : finset ℕ), simp,
+  rw hdivs at memy, simp only [finset.mem_insert, finset.mem_singleton] at memy,
+  cases memy, exfalso, apply hneq, symmetry, apply memy,
+  apply memy
+end
+
+lemma coprime_factor_eq_gcd_left {a b m n : ℕ} (cop : m.coprime n) (am : a ∣ m) (bn : b ∣ n):
+a = (a * b).gcd m :=
+begin
+  rw nat.gcd_eq_left_iff_dvd at am,
+  conv_lhs {rw ← am}, symmetry,
+  apply nat.coprime.gcd_mul_right_cancel a,
+  apply nat.coprime.coprime_dvd_left bn cop.symm,
+end
+
+lemma sum_divisors_multiplicative (m n : ℕ) :
+nat.coprime m n → sum_divisors (m * n) = (sum_divisors m) * (sum_divisors n) :=
+begin
+  intro cop,
+  by_cases m = 0, { rw h, simp }, have mne0 := h,
+  by_cases n = 0, { rw h, simp }, have nne0 := h,
+  repeat {rw sum_divisors},
+  rw finset.sum_mul_sum,
+  symmetry,
+  apply finset.sum_bij (λ x : ℕ × ℕ, λ (h : x ∈ _), x.fst * x.snd),
+  { simp only [prod.forall], intros a b hab, dsimp, simp only [finset.mem_product] at hab,
+    repeat {rw mem_divisors at hab}, rw mem_divisors,
+    split, apply mul_dvd_mul hab.left.left hab.right.left, 
+    intro contra, cases nat.eq_zero_of_mul_eq_zero contra,
+    apply hab.left.right h_1, apply hab.right.right h_1 },
+  { simp only [prod.forall], intros a b hab, dsimp, simp only [finset.mem_product] at hab },
+  { simp only [prod.forall], intros a b c hab h1 h2, 
+    dsimp at *, rw finset.mem_product at *, repeat {rw mem_divisors at *},
+    ext; dsimp,
+    { transitivity (a * b).gcd m,
+      { apply coprime_factor_eq_gcd_left cop hab.left.left hab.right.left },
+      { rw h2, symmetry, apply coprime_factor_eq_gcd_left cop h1.left.left h1.right.left } },
+    { transitivity (b * a).gcd n,
+      { apply coprime_factor_eq_gcd_left cop.symm hab.right.left hab.left.left },
+      { rw mul_comm, rw h2, rw mul_comm, symmetry,
+      apply coprime_factor_eq_gcd_left cop.symm h1.right.left h1.left.left } } },
+  { simp only [exists_prop, prod.exists, finset.mem_product],
+    intros c hc,
+    existsi c.gcd m, existsi c.gcd n, split,
+    { split; rw mem_divisors; split, apply nat.gcd_dvd_right _, apply mne0,
+      apply nat.gcd_dvd_right _, apply nne0 },
+    { rw ← nat.coprime.gcd_mul c cop,
+      symmetry, apply nat.gcd_eq_left,
+      rw mem_divisors at hc, apply hc.left } }
+end
+
+end basic_lemmas
+
+section mersenne_to_perfect
+
+theorem two_le_exponent_of_prime_mersenne {p : ℕ} :
+nat.prime (2 ^ p - 1) → 2 ≤ p :=
+begin
+  intro np,
+  cases p,
+  { simp only [nat.sub_self, nat.pow_zero] at np,
+    exfalso, apply nat.not_prime_zero np },
+  cases p, 
+  { simp only [nat.succ_sub_succ_eq_sub, nat.sub_zero, nat.pow_one] at np,
+    exfalso, apply nat.not_prime_one np },
+  omega,
+end
+
+/-
```

Comment:
What is your plan with this piece of code?

Reply:
I deleted it.